File: solutions/Leetcode_204/leetcode_204.cc
#include <math.h>
#include <numeric>
#include <vector>

using namespace std;

class Solution1
{
public:
    int countPrimes(int n)
    {
        if (n < 3)
            return 0;
        int n_sqrt = sqrt(n);
        vector<int> is_prime(n, 1);
        is_prime[0] = 0;
        is_prime[1] = 0;
        for (int x = 2; x <= n_sqrt; ++x)
        {
            if (is_prime[x])
            {
                for (int y = x; x * y < n; y++)
                    is_prime[x * y] = 0;
            }
        }
        return accumulate(is_prime.begin(), is_prime.end(), 0);
    }
};

File: solutions/Leetcode_204/leetcode_204.py
import math


class Solution1:
    def count_primes


class SolutionMINE:
    def count_primes(self, n: int) -> int:
        if n < 3:
            return 0
        sqrt_n = math.floor(math.sqrt(n))
        is_prime = [0, 0] + [1 for _ in range(n-2)]
        for x in range(2, sqrt_n+1):
            if is_prime[x]:
                is_prime[x*x:n:x] = [0] * len(is_prime[x*x:n:x])
        return sum(is_prime)


if __name__ == '__main__':
    n = 2
    ans = SolutionMINE().count_primes(n)
    print(ans)
